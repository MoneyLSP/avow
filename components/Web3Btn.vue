<template>
  <div>
    <v-btn role="button" @click.p="connectToWeb3">
      <v-icon>mdi-plus</v-icon>
      {{ btnAction }}
    </v-btn>
  </div>
</template>
<script lang="ts">
import { Component, namespace, Vue } from "nuxt-property-decorator"

const web3 = namespace("web3")
const contracts = namespace("contracts")

@Component
export default class Web3Btn extends Vue {
  @web3.State
  isConnected!: boolean

  @web3.Action
  connectWeb3!: () => Promise<void>

  @web3.Action
  registerListeners!: () => Promise<void>

  @web3.Mutation
  clearProvider!: () => void

  @contracts.Action
  initializeContracts!: () => Promise<void>

  @contracts.Action
  updateTokenBalances!: () => Promise<void>

  get btnAction() {
    return this.isConnected ? "Disconnect" : "Connect"
  }

  async connect() {
    await this.connectWeb3()
    await this.registerListeners()
    await this.initializeContracts()
    await this.updateTokenBalances()
  }

  clear() {
    this.clearProvider()
  }

  async connectToWeb3() {
    this.isConnected ? this.clear() : await this.connect()
  }
}
</script>
