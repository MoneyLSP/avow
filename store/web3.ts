import {
  Module,
  VuexAction as Action,
  VuexModule,
  VuexMutation as Mutation,
} from "nuxt-property-decorator"
import { ethers } from "ethers"
import { MetaMaskInpageProvider } from "@metamask/providers"

let currentProvider: ethers.providers.Web3Provider | undefined

export const getCurrentProvider = () => {
  return currentProvider
}

@Module({
  stateFactory: true,
  name: "web3",
  namespaced: true,
})
export default class web3 extends VuexModule {
  isConnected = false
  modalInitializing = false
  providerSet = false
  networkInfo: ethers.providers.Network = { name: "", chainId: -1 }
  correctNetwork = "kovan"
  activeNetwork = ""
  selectedAccount = ""
  blockTimestamp = 0

  get onCorrectNetwork() {
    return this.networkInfo.name === this.correctNetwork
  }

  get signer() {
    const modalProvider = getCurrentProvider()
    if (modalProvider !== undefined) {
      return modalProvider.getSigner(this.selectedAccount)
    } else {
      return undefined
    }
  }

  @Mutation
  setNetworkInfo(networkInfo: ethers.providers.Network) {
    const provider: unknown = getCurrentProvider()?.provider
    const metamaskProvider = provider as MetaMaskInpageProvider
    this.selectedAccount = metamaskProvider.selectedAddress ?? ""
    this.networkInfo = networkInfo
    this.activeNetwork = networkInfo.name
  }

  @Mutation
  setSelectedAccount(address: string) {
    this.selectedAccount = address
  }

  @Mutation
  setConnectionStatus(status: boolean) {
    this.isConnected = status
  }

  @Mutation
  setEthersProvider(provider: MetaMaskInpageProvider) {
    try {
      currentProvider = new ethers.providers.Web3Provider(provider as never)
      this.providerSet = true
    } catch (e: unknown) {
      console.log(
        "error converting to web3Provider to ethersProvider: %s",
        provider
      )
      this.providerSet = false
    }
  }

  @Mutation
  setModalInitializing(state: boolean) {
    this.modalInitializing = state
  }

  @Mutation
  clearProvider() {
    currentProvider = undefined
    this.providerSet = false
    this.isConnected = false
  }

  @Mutation
  setBlockTimestamp(timestamp: number) {
    this.blockTimestamp = timestamp
  }

  @Action({ rawError: true })
  registerListeners(metamaskProvider: MetaMaskInpageProvider) {
    if (metamaskProvider.isMetaMask) {
      console.log("Registering account listener")
      metamaskProvider.on("accountsChanged", async (accounts) => {
        const account = (accounts as Array<string>)[0]
        console.log("Detected account update: %s", account)
        this.context.commit("setSelectedAccount", account)
        await this.context.dispatch(
          "contracts/updateContractData",
          {},
          { root: true }
        )
      })

      // Note that this will not be triggered if we change between networks with the same chain id
      // Tried this but did not work: https://docs.ethers.io/v5/concepts/best-practices/#best-practices
      metamaskProvider.on("chainChanged", () => {
        console.log("Detected network change, reload page")
        window.location.reload()
      })
    }
    const ethersProvider = new ethers.providers.Web3Provider(
      metamaskProvider as any
    )
    ethersProvider.on("block", async (blockNumber) => {
      const block = await ethersProvider.getBlock(blockNumber)
      console.log("Block timestamp:", block.timestamp)
      this.context.commit("setBlockTimestamp", block.timestamp)
    })
  }

  @Action({ rawError: true })
  async connectWeb3() {
    this.context.commit("setModalInitializing", true)
    const provider = (window as any).ethereum as MetaMaskInpageProvider
    try {
      this.context.commit("setEthersProvider", provider)
      this.context.commit("setConnectionStatus", true)
      await this.context.dispatch("registerListeners", provider)
      await this.context.dispatch("updateNetworkInfo")
    } catch (e: unknown) {
      console.log("Error connecting to Web3")
      this.context.commit("setConnectionStatus", false)
    } finally {
      this.context.commit("setModalInitializing", false)
    }
  }

  @Action({ rawError: true })
  async updateNetworkInfo() {
    const ethersWeb3Provider = getCurrentProvider()
    if (this.providerSet && ethersWeb3Provider) {
      const networkInfo = await ethersWeb3Provider.getNetwork()
      this.context.commit("setNetworkInfo", networkInfo)
    }
  }
}
