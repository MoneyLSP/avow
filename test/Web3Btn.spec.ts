import Vuetify from "vuetify"
import { createLocalVue, mount, MountOptions } from "@vue/test-utils"
import { Store } from "vuex"
import { getModule } from "nuxt-property-decorator"
import Vue from "vue"
import Web3Btn from "~/components/Web3Btn.vue"
import { createStore } from "~/test/setup"
import web3 from "~/store/web3"

describe("Web3Btn", () => {
  const localVue = createLocalVue()
  let vuetify: any
  let store: Store<any>

  beforeEach(() => {
    vuetify = new Vuetify()
    store = createStore({ modules: { web3 } })
  })
  const mountFunction = (options: MountOptions<Vue>) => {
    return mount(Web3Btn, {
      localVue,
      vuetify,
      store,
      ...options,
    })
  }

  it("should mount properly", () => {
    const wrapper = mountFunction({})
    expect(wrapper.exists()).toBe(true)
  })

  it("should match snapshot", () => {
    const wrapper = mountFunction({})
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should try to initialize the ", async () => {
    // Todo mock connection to web3
    const wrapper = mountFunction({})
    const btn = wrapper.find('[role="button"]')
    const web3Store = getModule(web3, store)
    await btn.trigger("click")
    expect(web3Store.modalInitializing).toBe(true)
  })
})
