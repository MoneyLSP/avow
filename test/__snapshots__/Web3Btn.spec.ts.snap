// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Web3Btn should match a snapshot 1`] = `
<button
  class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default"
  role="button"
  type="button"
>
  <span
    class="v-btn__content"
  >
    <i
      aria-hidden="true"
      class="v-icon notranslate mdi mdi-plus theme--light"
    />
    
  Connect

  </span>
</button>
`;
